### PR TITLE
Only perform actions on the rbd pool after it has been created

### PR DESCRIPTION
**Pull request by AlLau**
The rbd pool is the default pool that gets created during ceph cluster
initializaiton.  If we act on the rbd related operations too early, the
rbd pool does not exist yet.  Move the call to perform rbd operations
to a later stage after other pools have been created.

The rbd_pool.yml playbook has all the operations related to the rbd pool.

Replace the always_run (deprecated) directive with check_mode.

Most of the ceph related tasks only need to run once.  The run_once directive
executes the task on the first host.

The ceph sub-command to delete a pool is delete (not rm).

The changes submitted here were tested with this ceph version.
ceph version 0.94.9-9.el7cp (b83334e01379f267fb2f9ce729d74a0a8fa1e92c)

### CODE REVIEW

**Comment by AlLau** [comment]
I looked at the console output for "Testing: playbook syntax" at https://2.jenkins.ceph.com/job/ceph-ansible-pr-syntax-check/751/console

It is complaining about this line.  Am I reading this correctly?

"
The error appears to have been in '/home/jenkins-build/build/workspace/ceph-ansible-pr-syntax-check/ceph-ansible/roles/ceph-rgw/tasks/multisite/master.yml': line 6, column 35, but may
be elsewhere in the file depending on the exact syntax problem.

The offending line appears to be:

  when:
    - 'No such file or directory' in realmcheck.stderr
                                  ^ here
"

**Comment by AlLau** [comment]
On the "Testing: luminous-*" failures, look like this task in ceph_keys.yml is complaining about the initial_mon_keyring.  The initial_mon_keyring is set in roles/ceph-common/tasks/facts_mon_fsid.yml, I think.

- name: test if initial mon keyring is in mon kv store
  command: ceph --cluster {{ cluster }} config-key get initial_mon_keyring
  changed_when: false
  ignore_errors: true
  always_run: true
  run_once: true
  failed_when: false
  register: is_initial_mon_keyring_in_kv

File: roles/ceph-mon/tasks/ceph_keys.yml [hunk 1]
```diff
@@ -86,47 +86,19 @@
...
   when:
     - openstack_config
     - inventory_hostname == groups[mon_group_name] | last
 
+# CEPH creates the rbd pool during the ceph cluster initialization.  If
+# the rbd_pool.yml playbook is called too early, the rbd pool does not
+# exist yet.
+- name: include rbd_pool.yml
+  include: rbd_pool.yml
+
```
Comment (leseb) [inline roles/ceph-mon/tasks/ceph_keys.yml]: please add this directive:

`ceph_release_num.{{ ceph_release }} < ceph_release_num.luminous`

The `rbd` is not created by default in and after Luminous.

File: roles/ceph-mon/tasks/rbd_pool_df.yml [hunk 2]
```diff
@@ -0,0 +1,14 @@
+---
+
```
Comment (leseb) [inline roles/ceph-mon/tasks/rbd_pool_df.yml]: please remove this empty line

File: roles/ceph-mon/tasks/rbd_pool_df.yml [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+- assert:
```
Comment (leseb) [inline roles/ceph-mon/tasks/rbd_pool_df.yml]: We normally use the `fail` module in the playbook, please use it instead

File: roles/ceph-mon/tasks/rbd_pool_luminous.yml [hunk 4]
```diff
@@ -0,0 +1,11 @@
+---
+
+- assert:
```
Comment (leseb) [inline roles/ceph-mon/tasks/rbd_pool_luminous.yml]: ditto

File: roles/ceph-mon/tasks/rbd_pool_size.yml [hunk 5]
```diff
@@ -1,16 +1,23 @@
 ---
+
+- assert:
```
Comment (leseb) [inline roles/ceph-mon/tasks/rbd_pool_size.yml]: ditto

File: roles/ceph-mon/tasks/rbd_pool_pgs.yml [hunk 6]
```diff
@@ -1,10 +1,25 @@
 ---
+
+- assert:
```
Comment (leseb) [inline roles/ceph-mon/tasks/rbd_pool_pgs.yml]: ditto

File: roles/ceph-mon/tasks/rbd_pool_luminous.yml [hunk 7]
```diff
@@ -0,0 +1,11 @@
...
+    that: rbd_pool_exist.rc != 0
+    msg: "Expect the rbd pool not exist in rbd_pool_luminous"
+
+- name: create rbd pool on luminous
+  shell: ceph --connect-timeout 5 --cluster {{ cluster }} osd pool create rbd {{ ceph_conf_overrides.global.osd_pool_default_pg_num }}
```
Comment (leseb) [inline roles/ceph-mon/tasks/rbd_pool_luminous.yml]: please use the `command` module

File: roles/ceph-mon/tasks/rbd_pool.yml [hunk 8]
```diff
@@ -1,17 +1,36 @@
...
+    - ceph_conf_overrides.global.osd_pool_default_pg_num is defined
+
+- name: include rbd_pool_size.yml
+  include: rbd_pool_size.yml
+  when:
+    - rbd_pool_exist.rc == 0
+    - global_in_ceph_conf_overrides
+    - ceph_conf_overrides.global.osd_pool_default_size is defined
 
+- name: include rbd_pool_luminous.yml
```
Comment (leseb) [inline roles/ceph-mon/tasks/rbd_pool.yml]: I personally won't change the default behavior of Ceph, so for and after Luminous let's not creating the RBD pool.

**Comment by leseb** [comment]
@AlLau don't bother with the `playbook syntax`  for now.

File: roles/ceph-mon/tasks/rbd_pool_df.yml [hunk 9]
```diff
@@ -0,0 +1,14 @@
+---
+- name: Verify that rbd pool exist
```
Comment (leseb) [inline roles/ceph-mon/tasks/rbd_pool_df.yml]: please remove any capital letters from the name of the task

File: roles/ceph-mon/tasks/rbd_pool_pgs.yml [hunk 10]
```diff
@@ -1,10 +1,25 @@
 ---
+- name: Verify that rbd pool exist
```
Comment (leseb) [inline roles/ceph-mon/tasks/rbd_pool_pgs.yml]: ditto

File: roles/ceph-mon/tasks/rbd_pool_size.yml [hunk 11]
```diff
@@ -1,16 +1,23 @@
 ---
+- name: Verify that rbd pool exist
```
Comment (leseb) [inline roles/ceph-mon/tasks/rbd_pool_size.yml]: ditto

File: roles/ceph-mon/tasks/ceph_keys.yml [hunk 12]
```diff
@@ -86,47 +86,20 @@
...
-    - rbd_pool_exist.rc != 0
-
+# Create the pools listed in openstack_pools
 - name: include openstack_config.yml
   include: openstack_config.yml
   when:
     - openstack_config
     - inventory_hostname == groups[mon_group_name] | last
 
+# CEPH creates the rbd pool during the ceph cluster initialization in
```
Comment (guits) [inline roles/ceph-mon/tasks/ceph_keys.yml]: please, remove the trailing whitespace at this line.

**Comment by guits** [comment]
The error in the CI are not related to this PR.

**Comment by AlLau** [comment]
I don't think I have write access to ceph-ansible.  Can someone help merge this pull request?  Thanks.

**Comment by guits** [comment]
@AlLau your commit will be merged in #1976.
I opened this PR with your commit (no worries, you are still the author of course)
I did this because I squashed your 3 commits and need to add other commits related to your initial PR.

**Pull request closed without merging**